pallets/collator-selection/src/lib.rs

```diff
@@ -499,29 +512,83 @@ pub mod pallet {
 			collators.extend(candidates.into_iter().collect::<Vec<_>>());
 			collators
 		}
-		/// Kicks out and candidates that did not produce a block in the kick threshold.
+
+		/// Removes collators with unsatisfactory performance
+		/// Returns the removed AccountIds
 		pub fn kick_stale_candidates(
 			candidates: Vec<CandidateInfo<T::AccountId, BalanceOf<T>>>,
 		) -> Vec<T::AccountId> {
-			let now = frame_system::Pallet::<T>::block_number();
-			let kick_threshold = T::KickThreshold::get();
-			candidates
+			// 0. TODO: All sanity checks
+			let mut collator_perf_this_session =
+				<BlocksPerCollatorThisSession<T>>::iter().collect::<Vec<_>>();
+			if collator_perf_this_session.is_empty() {
+				return Vec::new();
+			}
+			// 1. Sort collator performance list
+			collator_perf_this_session.sort_unstable_by_key(|k| k.1); // XXX: don't like the tuple accessor, could this be a struct?
+														  // collator_perf_this_session.reverse();
+			let no_of_candidates = collator_perf_this_session.len();
+
+			// 2. get percentile by _exclusive_ nearest rank method https://en.wikipedia.org/wiki/Percentile#The_nearest-rank_method (rust percentile API is feature gated)
+			// RAD: Doublecheck that f64 as usize actually rounds up
+			let ordinal_rank = (((T::PerformancePercentileToConsiderForKick::get() as f64) / 100.0
+				* no_of_candidates as f64) as usize)
+				.saturating_sub(1); // Note: -1 to accomodate 0-index counting // RAD: Is there API to saturate but notify on non-overflow so we can log a warning?
```

Comment:
Is there API to saturate but notify on a prevented overflow so we can log a warning?

Reply:
Suggested way is something like the following:

```rust
let consider_for_kick_percentile = T::PerformancePercentileToConsiderForKick::get();
let ordinal_rank = match ((no_of_candidates * (consider_for_kick_percentile as f64) / 100.0)
    .ceil() as usize)
    .saturating_sub(1)
{
    Some(ordinal_rank) => ordinal_rank,
    _ => {
        log::info!(...);
        0
    }
};
```

and then you can use `consider_for_kick_percentile` later on [line 550](https://github.com/Manta-Network/Manta/pull/358/files#diff-a0359a0add814d66aac74493f7e17aefdcee21d3de1ae592e97059082b6feedfR550) instead of doing another database GET.

Reply:
thanks, I added your  `consider_for_kick_percentile` suggestion.

Turns out we don't need checked_sub because the saturation case can not happen (due to new precondition check)